[JSC] Fix Date Time String Parsing

Gumichocopengin8 opened the pull request:
#### 97b469eda5e1abbad820563e16edbe4aba147454
<pre>
[JSC] Fix Date Time String Parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=294600">https://bugs.webkit.org/show_bug.cgi?id=294600</a>

Reviewed by Ross Kirsling.

This patch fixed the Date String Format parsing to align with the TC39 spec [1].
If the string format conforms to the Date Time String Format (ISO8601),
it is parsed according to the spec [2].

[1]: <a href="https://tc39.es/ecma262/#sec-date.parse">https://tc39.es/ecma262/#sec-date.parse</a>
[2]: <a href="https://tc39.es/ecma262/#sec-date-time-string-format">https://tc39.es/ecma262/#sec-date-time-string-format</a>

* JSTests/stress/relaxed-timezone-designators.js:
* JSTests/test262/expectations.yaml:
* LayoutTests/js/date-parse-test-expected.txt:
* LayoutTests/js/script-tests/date-parse-test.js:
* Source/WTF/wtf/DateMath.cpp:
(WTF::parseES5DatePortion):
(WTF::parseES5TimePortion):
(WTF::parseES5Date):

Canonical link: <a href="https://commits.webkit.org/296339@main">https://commits.webkit.org/296339@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7cc9f117945a93745d19466fcf78ecd1f947fd3a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108217 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27879 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18300 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113427 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28577 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36430 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82166 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111165 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22645 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97481 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62598 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22060 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15617 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58162 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/100804 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92011 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15673 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116551 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/106777 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35280 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25976 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91192 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35653 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93757 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90987 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23189 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35876 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13642 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31058 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Failed to compile WebKit; Compiled WebKit (cancelled)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35180 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40732 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131064 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34912 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35576 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38266 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36580 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Gumichocopengin8 commented on JSTests/stress/relaxed-timezone-designators.js:
Test cases were incorrect.

rkirsling commented on Source/WTF/wtf/DateMath.cpp:
```diff
@@ -471,7 +471,9 @@ static bool parseES5DatePortion(std::span<const LChar>& currentPosition, int& ye
-    if ((postParsePosition.data() - currentPosition.data()) != 2)
+    if ((postParsePosition.data() - currentPosition.data()) == 1)
+        isSingleDigit = true;
+    if ((postParsePosition.data() - currentPosition.data()) != 2 && (postParsePosition.data() - currentPosition.data()) != 1)
```
This should just be an `else if`.

Gumichocopengin8 replied:
Thanks for the review. Fixed.

rkirsling replied:
Er, I mean that you're duplicating the condition unnecessarily. The `else if` replaces `(postParsePosition.data() - currentPosition.data()) != 1`.

Gumichocopengin8 replied:
Ah, thanks for pointing this out. Fixed!

rkirsling commented on Source/WTF/wtf/DateMath.cpp:
```diff
@@ -484,7 +486,9 @@ static bool parseES5DatePortion(std::span<const LChar>& currentPosition, int& ye
-    if ((postParsePosition.data() - currentPosition.data()) != 2)
+    if ((postParsePosition.data() - currentPosition.data()) == 1)
+        isSingleDigit = true;
+    if ((postParsePosition.data() - currentPosition.data()) != 2 && (postParsePosition.data() - currentPosition.data()) != 1)
```
Ditto.

rkirsling reviewed (approved):
Having verified that this matches SM/V8 behavior, r=me with comments.

Gumichocopengin8 committed:
[JSC] Fix Date Time String Parsing
https://bugs.webkit.org/show_bug.cgi?id=294600

Reviewed by Ross Kirsling.

This patch fixed the Date String Format parsing to align with the TC39 spec [1].
If the string format conforms to the Date Time String Format (ISO8601),
it is parsed according to the spec [2].

[1]: https://tc39.es/ecma262/#sec-date.parse
[2]: https://tc39.es/ecma262/#sec-date-time-string-format

* JSTests/stress/relaxed-timezone-designators.js:
* JSTests/test262/expectations.yaml:
* LayoutTests/js/date-parse-test-expected.txt:
* LayoutTests/js/script-tests/date-parse-test.js:
* Source/WTF/wtf/DateMath.cpp:
(WTF::parseES5DatePortion):
(WTF::parseES5TimePortion):
(WTF::parseES5Date):

Canonical link: https://commits.webkit.org/296339@main
